pkg/agent/proxy/proxier.go

```diff
@@ -136,65 +136,65 @@ func (p *proxier) isInitialized() bool {
 	return p.endpointsChanges.Synced() && p.serviceChanges.Synced()
 }
 
-// removeStaleServices removes all expired Services. Once a Service is deleted, all
-// its Endpoints will be expired, and the removeStaleEndpoints method takes
-// responsibility for cleaning up, thus we don't need to call removeEndpoint in this
-// function.
+// removeStaleServices removes all the configurations of expired Services and their associated Endpoints.
 func (p *proxier) removeStaleServices() {
 	for svcPortName, svcPort := range p.serviceInstalledMap {
 		if _, ok := p.serviceMap[svcPortName]; ok {
 			continue
 		}
 		svcInfo := svcPort.(*types.ServiceInfo)
-		klog.V(2).Infof("Removing stale Service: %s %s", svcPortName.Name, svcInfo.String())
+		svcInfoStr := svcInfo.String()
+		klog.V(2).InfoS("Removing stale Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 		if err := p.ofClient.UninstallServiceFlows(svcInfo.ClusterIP(), uint16(svcInfo.Port()), svcInfo.OFProtocol); err != nil {
-			klog.ErrorS(err, "Failed to remove flows of Service", "Service", svcPortName)
+			klog.ErrorS(err, "Error when uninstalling ClusterIP flows for Service", "ServiceName", svcPortName, "ServiceInfo", svcInfoStr)
 			continue
 		}
-
-		if p.proxyAll {
-			// Remove NodePort flows and configurations.
-			if svcInfo.NodePort() > 0 {
-				if err := p.uninstallNodePortService(uint16(svcInfo.NodePort()), svcInfo.OFProtocol); err != nil {
-					klog.ErrorS(err, "Failed to remove flows and configurations of Service", "Service", svcPortName)
-					continue
-				}
+		// Remove associated Endpoints flows.
+		if endpoints, ok := p.endpointsInstalledMap[svcPortName]; ok {
+			if err := p.removeStaleEndpoints(endpoints, svcInfo.Protocol()); err != nil {
```

Comment:
I don't find where it removes cache in `endpointsInstalledMap` in this case. Seems an unit test is needed.

Reply:
Added `delete(p.endpointsInstalledMap, svcPortName)` after err is nil and related unit tests.